File: worker/geoip.ts
import dns from 'dns';
import geoip from 'geoip-country';

const ipv4Regex = /^(25[0-5]|2[0-4][0-9]|[0-1]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[0-1]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[0-1]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[0-1]?[0-9][0-9]?)$/;
const ipv6Regex = /^([0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}$|^([0-9a-fA-F]{1,4}:){1,7}:$|^([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}$|^([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2}$|^([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}$|^([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}$|^([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}$|^[0-9a-fA-F]{1,4}(:[0-9a-fA-F]{1,4}){1,6}$|^:(:[0-9a-fA-F]{1,4}){1,7}$|^::([0-9a-fA-F]{1,4}){1,6}$|^([0-9a-fA-F]{1,4}:){1,4}::([0-9a-fA-F]{1,4}){1,4}$|^::([0-9a-fA-F]{1,4}){1,7}$|^::([0-9a-fA-F]{1,4}){1,6}$|^[0-9a-fA-F]{1,4}::([0-9a-fA-F]{1,4}){1,5}$|^([0-9a-fA-F]{1,4}){1,4}::([0-9a-fA-F]{1,4}){1,5}$|^::([0-9a-fA-F]{1,4}){1,6}$|^[0-9a-fA-F]{1,4}::([0-9a-fA-F]{1,4}){1,4}$|^([0-9a-fA-F]{1,4}:){1,6}::([0-9a-fA-F]{1,4}){1,5}$|^([0-9a-fA-F]{1,4}:){1,5}::([0-9a-fA-F]{1,4}){1,6}$|^[0-9a-fA-F]{1,4}::([0-9a-fA-F]{1,4}){1,4}$|^([0-9a-fA-F]{1,4}:){1,4}::([0-9a-fA-F]{1,4}){1,4}$/i;
const isIPAddress = function (ip: string): boolean {
    return ipv4Regex.test(ip) || ipv6Regex.test(ip);
};

const resolve = function (domain: string): Promise<string> {
    return new Promise((resolve, reject) => {
        dns.lookup(domain, (err, address) => {
            if (err) {
                reject(err);
            } else {
                resolve(address);
            }
        });
    });
}

export const getCountry = async function (domainOrIp: string): Promise<string> {
    if (!isIPAddress(domainOrIp)) {
        domainOrIp = await resolve(domainOrIp)
    }
    return await geoip.lookup(domainOrIp)?.country;
};

File: validator/src/geoip.rs
use maxminddb;
use std::error::Error;
use std::net::ToSocketAddrs;
use std::net::{IpAddr, Ipv4Addr, Ipv6Addr};
use std::str::FromStr;

pub type GeoIpClient = maxminddb::Reader<Vec<u8>>;

fn is_ipv4(ip: &str) -> bool {
    ip.parse::<std::net::Ipv4Addr>().is_ok()
}

fn is_ipv6(ip: &str) -> bool {
    ip.parse::<std::net::Ipv6Addr>().is_ok()
}

fn is_ip_address(ip: &str) -> bool {
    is_ipv4(ip) || is_ipv6(ip)
}

fn resolve(domain: &str) -> Result<IpAddr, Box<dyn Error>> {
    let mut addrs = (domain, 0).to_socket_addrs()?;
    if let Some(addr) = addrs.next() {
        Ok(addr.ip())
    } else {
        Err("No valid IP address found".into())
    }
}

fn str_to_ip(ip: &str) -> Result<IpAddr, Box<dyn Error>> {
    if is_ipv4(ip) {
        Ok(IpAddr::V4(Ipv4Addr::from_str(ip)?))
    } else if is_ipv6(ip) {
        Ok(IpAddr::V6(Ipv6Addr::from_str(ip)?))
    } else {
        Err("Invalid IP address".into())
    }
}

pub struct GeoIp {
    reader: GeoIpClient,
}

impl GeoIp {
    pub fn new(path: &str) -> Result<Self, Box<dyn Error>> {
        Ok(Self {
            reader: maxminddb::Reader::open_readfile(path)?,
        })
    }

    pub fn get_country(&self, ip_or_domain: &str) -> Result<String, Box<dyn Error>> {
        if ip_or_domain.ends_with(".onion") {
            return Ok("TOR".to_string());
        }

        let ip: IpAddr = if is_ip_address(ip_or_domain) {
            str_to_ip(ip_or_domain)?
        } else {
            resolve(ip_or_domain)?
        };

        let result = self.reader.lookup(ip)?;
        
        let country: maxminddb::geoip2::Country = result
            .decode()?
            .ok_or("Country information could not be found")?;

        match country.country.iso_code {
            Some(code) => Ok(code.to_string()),
            None => Err("No country code found".into()),
        }
    }
}
